tests/unit/test_compress.py

```diff
@@ -0,0 +1,276 @@
+# Copyright 2023 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"). You
+# may not use this file except in compliance with the License. A copy of
+# the License is located at
+#
+# http://aws.amazon.com/apache2.0/
+#
+# or in the "license" file accompanying this file. This file is
+# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+import gzip
+import io
+
+import pytest
+
+from botocore.compress import maybe_compress_request
+from botocore.config import Config
+from tests import mock
+
+
+def _op_with_compression():
+    op = mock.Mock()
+    op.request_compression = {'encodings': ['gzip']}
+    op.has_streaming_input = False
+    return op
+
+
+def _op_with_multiple_compressions():
+    op = mock.Mock()
+    # TODO: Update second value to a different encoding. Currently only gzip
+    # is supported.
+    op.request_compression = {'encodings': ['gzip', 'gzip']}
+    op.has_streaming_input = False
+    return op
+
+
+def _op_unknown_compression():
+    op = mock.Mock()
+    op.request_compression = {'encodings': ['foo']}
+    op.has_streaming_input = None
+    return op
+
+
+def _op_without_compression():
+    op = mock.Mock()
+    op.request_compression = None
+    op.has_streaming_input = False
+    return op
+
+
+def _streaming_op_with_compression():
+    op = _op_with_compression()
+    op.has_streaming_input = True
+    streaming_shape = mock.Mock()
+    streaming_shape.metadata = {}
+    op.get_streaming_input.return_value = streaming_shape
+    return op
+
+
+def _streaming_op_with_compression_requires_length():
+    op = _streaming_op_with_compression()
+    streaming_shape = mock.Mock()
+    streaming_shape.metadata = {'requiresLength': True}
+    op.get_streaming_input.return_value = streaming_shape
+    return op
+
+
+OP_WITH_COMPRESSION = _op_with_compression()
+OP_UNKNOWN_COMPRESSION = _op_unknown_compression()
+STREAMING_OP_WITH_COMPRESSION = _streaming_op_with_compression()
```

Comment:
These feel a bit unnecessary, especially since you're assigning them all to constants. Could we simplify this with something like:

```suggestion

def _make_op(request_compression=None, streaming=None, metadata=None):
    op = mock.Mock()
    op.request_compression = request_compression
    op.has_streaming_input = streaming_input
    if metadata is not None:
        streaming_shape = mock.Mock()
        streaming_shape.metadata = metadata
        op.get_streaming_input.return_value = streaming_shape
    return op


OP_WITH_COMPRESSION = _make_op({'encodings': ['gzip']}, False)
OP_WITHOUT_COMPRESSION = _make_op(None, False)
OP_WITH_MULTIPLE_COMPRESSION = _make_op({'encodings': ['gzip', 'gzip']}, False)
OP_UNKNOWN_COMPRESSION = _make_op({'encodings': ['foo']}, None)
STREAMING_OP_WITH_COMPRESSION = _make_op({'encodings': ['gzip']}, True, {})
STREAMING_OP_WITH_REQUIRES_LENGTH = _make_op({'encodings': ['gzip']}, True, {'requiresLength': True})

```